client-core/src/service/wallet_state_service.rs

```diff
@@ -192,20 +187,19 @@ impl Default for WalletState {
 
 impl WalletState {
     /// Applies memento to wallet state
-    fn apply_memento(&mut self, memento: &WalletStateMemento) -> Result<()> {
+    pub fn apply_memento(&mut self, memento: &WalletStateMemento) {
         for operation in memento.0.iter() {
-            self.apply_memento_operation(operation)?
+            self.apply_memento_operation(operation);
         }
-
-        Ok(())
     }
 
     /// Applies a memento operation to wallet state
-    fn apply_memento_operation(&mut self, memento_operation: &MementoOperation) -> Result<()> {
+    fn apply_memento_operation(&mut self, memento_operation: &MementoOperation) {
         match memento_operation {
             MementoOperation::AddTransactionChange(ref transaction_id, ref transaction_change) => {
                 if !self.transaction_history.contains_key(transaction_id) {
-                    self.balance = (self.balance + transaction_change.balance_change)?;
+                    self.balance = (self.balance + transaction_change.balance_change)
+                        .expect("balance overflow");
```

Comment:
Handle error instead of `.expect()`

Reply:
I think `expect` here may be fine -- this would only fail if `self.balance + change` went over maximum total supply... is `transaction_change` validated or could it be random user input?

Reply:
if it's validated, expect is fine. If it's a random user input (e.g. balance_change could be `Coin::max()`), then Result make sense

Reply:
Right now, `AddTransactionChange` will only come from wallet synchroniser. But, I can see this functionality being used when user has the transaction data and manually enters (lite client) it instead of full-sync.

Reply:
`Coin` overflow is a phenomenon of a deep bug in the whole system, which I think we should probably just let it crash. Even with `-doverflow-check` on crate level in the future.

Reply:
> Right now, `AddTransactionChange` will only come from wallet synchroniser. But, I can see this functionality being used when user has the transaction data and manually enters (lite client) it instead of full-sync.

We should check user-input at the interface level.

Reply:
I don't think we should think of `Coin` overflow in this way. If this is the case, why not just panic at the source, i.e., in `Add` impl.

Reply:
> I don't think we should think of `Coin` overflow in this way. If this is the case, why not just panic at the source, i.e., in `Add` impl.

That is indeed a good option, `Add` panic, if you want `Option`, use `checked_add`.

Reply:
panicing at `Add impl.` of `Coin` would cause a network shutdown if there was a way to sneak in / gossip invalid transaction that causes an overflow, as all chain-abci would panic at the same time.

crashing a single wallet / client due to a bug (e.g. not validating its local user input) is a different story.

Reply:
I take back this `most of coin operations .. unwrap` statement, it probably just my wrong impression.
Talk about this case, if we are not sure if it's only used for validated block data, we might should check the overflow.
I guess we need some kind of guideline on this situation, or this discussion may happen again another time.